alf/algorithms/doac_algorithm.py

```diff
@@ -0,0 +1,263 @@
+                        epsilon_greedy=None,
+                        eps_greedy_sampling=False,
+                        rollout=False):
+        """
+        Differences between DSacAlgorithm._predict_action:
+
+        1. Only continuous actions are supported.
+
+        2. OAC explore policy is constructed from the target policy (actor_network) 
+        and used for action prediction during rollout.
+        """
+
+        # new_state = SacActionState()
+        # action_dist, actor_network_state = self._actor_network(
+        #     observation, state=state.actor_network)
+        # assert isinstance(action_dist, td.TransformedDistribution), (
+        #     "Squashed distribution is expected from actor_network.")
+        # assert isinstance(
+        #     action_dist.base_dist, dist_utils.DiagMultivariateNormal
+        # ), ("the base distribution should be diagonal multivariate normal.")
+        # normal_dist = action_dist.base_dist
+        # unsquashed_mean = normal_dist.mean
+        # unsquashed_std = normal_dist.stddev
+        # unsquashed_var = normal_dist.variance
+        # new_state = new_state._replace(actor_network=actor_network_state)
+
+        new_state = SacActionState()
+        action_dist, actor_network_state = self._actor_network(
+            observation, state=state.actor_network)
+        new_state = new_state._replace(actor_network=actor_network_state)
+
+        action_dist_type = nest.map_structure(
+            lambda dist: isinstance(dist, td.TransformedDistribution),
+            action_dist)
+        action_base_dist_type = nest.map_structure(
+            lambda dist: isinstance(dist.base_dist, dist_utils.
+                                    DiagMultivariateNormal), action_dist)
+        action_dist_type = nest.flatten(action_dist_type)
+        action_base_dist_type = nest.flatten(action_base_dist_type)
+        assert all(nest.flatten(action_dist_type)), (
+            "Squashed distribution is expected from actor_network.")
+        assert all(nest.flatten(action_base_dist_type)), (
+            "the base distribution should be diagonal multivariate normal.")
+
+        def mean_shift_fn(mu, dqda, sigma):
+            if self._dqda_clipping:
+                dqda = torch.clamp(dqda, -self._dqda_clipping,
+                                   self._dqda_clipping)
+            norm = torch.sqrt(torch.sum(torch.mul(dqda * dqda, sigma))) + 1e-6
+            shift = self._explore_delta * torch.mul(sigma, dqda) / norm
+            return mu + shift
+
+        if (rollout and not self._training_started):
+            # get batch size with ``get_outer_rank`` and ``get_nest_shape``
+            # since the observation can be a nest in the general case
+            outer_rank = nest_utils.get_outer_rank(observation,
+                                                   self._observation_spec)
+            outer_dims = alf.nest.get_nest_shape(observation)[:outer_rank]
```

Comment:
remembered that these appeared in other algorithms as well. maybe we can extract them as a common utility function so that it will be easier to be used in different places?

Reply:
Yes, similar codes are also used in ``oac_algorithm``. The mean_shift_fn depends on some member variables, so I leave it there. I find two other functions that could be extracted for both oac and doac.